### ISSUE TITLE
Weighted merge

### ISSUE THREAD

**Issue by gtuzi** [fchollet/keras#5095]
I am trying to compute a weighted output from multiple parallel models using Keras' Merge layer. I'm using Theano backend.

I have L parallel models (Ci). Each of their output layer is a k-sized softmax.

There is one model (N), its output is a L-sized softmax.

Here is what I have so far:

Parallel models (Ci) each with k dimension in the output layer:

    model.add(Dense(K, activation='softmax', W_regularizer=l2(0.001),init='normal'))

The weighing model (N), output layer:

    model.add(Dense(L, activation='softmax', W_regularizer=l2(0.001), init='normal'))

The merger is as follows:

    model.add(Merge(layers=model_group,
                    mode=lambda model_group: self.merge_fun(model_group, L),
                    output_shape = (None, k)))

where "model_group" is a (L+1)-length list [N, C1, C2, ..., CL], and merge_fun's signature is:

    def merge_fun(self, model_group, L):

Mathematically, I would like the output of the merged layer to be a weighted sum:

> *out = N[1]x([C11, C12, C13, .., C1k]) + N[2]x([C21, C22, C23, ..., C2k]) + ... + N[L]x([CL1, CL2, CL3, ..., CLk])*, 

where out is a vector of size k

How can I use the Merge layer to achieve this ? 

I know that the magic would probably have to happen in the 'merge_fun', but I am not sure how to perform matrix algebra in Keras. The tensor parameters don't have a "shape" parameter - they have a 'keras_shape' (None, K or L) - but I am not sure how to combine parallel models' output into a matrix. 

I tried using a local evaluation of the following expressions:

    K.concatenate([model_group[1], model_group[2]], axis=0)*model_group[0]

and 

    model_group[0] * K.concatenate([model_group[1], model_group[2]], axis=0)

both of which didn't throw an error, so I can't use this as a guide. After the multiplication, the result returned did not have the "keras_shape" variable, so I'm not sure what the shape of the result is.

Any suggestions ?

**Comment by bstriner** [comment]
When I'm trying to write a new layer, I make a new model which is just input layers and that layer, throw some np.random.random into it, and check if it runs and gives me what I expect. A lot of errors don't get thrown when the graph is built, only when it is run. Generate some random numbers for Ci and make L something like `to_categorical([0,1,2,3])`, and check if your output is coming from the right input.

You have the right idea, but you should be sending all of your layers to the merge.
```python
model.add(Merge(layers=model_group+[L],
                mode=lambda z: self.merge_fun(z[:-1], z[-1]),
                output_shape = lambda z:z[0]))
```

Also, note the output_shape. This takes the input shape from the first merged layer. When you run summary, this will output (None, k), but when it is running the model, it will be (batch_size, k).

The function is going to be a `K.sum(K.concatenate(...)*L)`. You will have to do some testing to see if you get all your axes right. You may have to use permute and dimshuffle to get everything to line up correctly.

BTW, you will have less trouble saving and loading your model if you move `merge_fun` out of your class and remove the `self` parameter. Otherwise, the model may not run correctly if you save and load it.

Cheers

**Comment by gtuzi** [comment]
Thanks Ben for the suggestions. I haven't had a chance to try them out yet (will soon here), but I think they're a good starting point. I will keep this opened for a bit longer (I'm sure I'll run into something).

**Comment by gsabran** [comment]
what is your `weighted_merge_fun` ? I've tried creating a new merge layer:
```python
class WeightedAverage(_Merge):
    """Layer that computes a mean between samples in two tensors according to weights computed in another vector.
    E.g. if applied to k tensors `a_0` and `a_k` of shape `(batch_size, n)`, and a tensor `b` of shape `(batch_size, k)`
    the output will be a tensor of shape `(batch_size, n)`
    where each entry `i` will be the dot product between
    `sum(a_j[i] * b[j], j)`.
    """
    def __init__(self, **kwargs):
        self.weight_tensor = None
        super(WeightedAverage, self).__init__(**kwargs)

    def _merge_function(self, inputs):
        if self.weight_tensor is None:
            raise ValueError('A `WeightedAverage` layer should have a weight layer')
        shape = inputs[0].get_shape().as_list()
        k = shape[1]
        n = len(inputs)
        all_inputs = K.reshape(Concatenate()(inputs), (-1, n, k))
        res = K.reshape(K.batch_dot(all_inputs, self.weight_tensor, axes=(1, 2)), (-1, k))
        return res

    def call(self, inputs, weights):
        shape = weights.get_shape().as_list()
        if shape[1] != len(inputs):
            raise ValueError('A `WeightedAverage` weights layer should have as many'
                             'outputs as there are inputs.')
        
        self.weight_tensor = K.reshape(weights, (-1, 1, shape[1]))
        return super(WeightedAverage, self).call(inputs)
```
But that's crashing at run time, even though the compilation phase works fine: https://stackoverflow.com/questions/44841463/keras-weighted-average-of-tensors I'm also not sure I'm getting the axis right

**Comment by HaythamAssem** [comment]
Can you please explain what is the "weighted_merge_fun" and any example for its implementation ?
